1-clients/src/main/scala/com/snowplowanalytics/piinguin/client/PiinguinClient.scala

```diff
@@ -0,0 +1,247 @@
+   * @return a future for the execution that returns either a failure message or a success message
+   */
+  def deletePiiRecord(modifiedValue: String): Future[Either[String, String]] = {
+    logger.debug(s"Sending single delete record request for modifiedValue: $modifiedValue")
+    stub.deletePiiRecord(DeletePiiRecordRequest(modifiedValue)).transform {
+      case Success(ChangeRecordResponse(None)) => errSuccess("Failed to get a response")
+      case Success(ChangeRecordResponse(Some(error))) =>
+        if (error.isSuccess) okSuccess(error.message) else errSuccess(error.message)
+      case Failure(t) => errSuccess(t.getMessage)
+    }
+  }
+
+  /**
+   * readPiiRecord reads a single piirecord using a single request asynchronously
+   * @param modifiedValue the modified value of the record
+   * @param basisForProcessing the legal basis for requesting to read this record
+   * @return a future for the execution that returns either a failure message or a pii record
+   */
+  def readPiiRecord(modifiedValue: String,
+                    basisForProcessing: LawfulBasisForProcessing): Future[Either[String, PiiRecord]] = {
+    logger.debug(
+      s"Sending single read record request for modifiedValue: $modifiedValue and basisForProcessing: $basisForProcessing")
+    stub.readPiiRecord(ReadPiiRecordRequest(modifiedValue, basisForProcessing)).transform {
+      case Success(ReadPiiRecordResponse(_, Some(error))) if (!error.isSuccess) => errSuccess(error.message)
+      case Success(ReadPiiRecordResponse(Some(piiRecord), _)) => Success(Right(piiRecord))
+      case Failure(t)                                         => errSuccess(t.getMessage)
+    }
+  }
+
+  /**
+   * createPiiRecordIO cats IO version of createPiiRecord call
+   */
+  def createPiiRecordIO(modifiedValue: String, originalValue: String): IO[Either[String, String]] = IO.fromFuture {
+    IO { createPiiRecord(modifiedValue, originalValue) }
+  }
+
+  /**
+   * deletePiiRecordIO cats IO version of deletePiiRecord
+   */
+  def deletePiiRecordIO(modifiedValue: String): IO[Either[String, String]] = IO.fromFuture {
+    IO { deletePiiRecord(modifiedValue) }
+  }
+
+  /**
+   * readPiiRecordIO cats IO version of readPiiRecord
+   */
+  def readPiiRecordIO(modifiedValue: String,
+                      basisForProcessing: LawfulBasisForProcessing): IO[Either[String, PiiRecord]] = IO.fromFuture {
+    IO { readPiiRecord(modifiedValue, basisForProcessing) }
+  }
+
+  /**
+   * EXPERIMENTAL
+   * Given two fs2 queues, it consumes requests from one and enqueues responses in the other.
+   *
+   * @param maxQueued the maximum requests to be queued before the request queue blocks
+   * @param kvPairs request queue
+   * @param responseQueue response queue
+   */
+  def createRecordsStreamingIO(maxQueued: Int,
+                               kvPairs: Queue[IO, (String, String)],
+                               responseQueue: Queue[IO, Either[String, String]]): IO[Unit] = {
```

Comment:
I'm assuming you haven't looked at how to provide a nicer api since our last talk?

Reply:
I have marked this as *HIGHLY EXPERIMENTAL* and will look into it in v2 (as I did not have time for it this time round).

Reply:
cool, I logged https://github.com/snowplow-incubator/piinguin/issues/11